### PR TITLE
Macos compile

### PR DESCRIPTION
**Pull request by koshelev**
Should allow to compile project without ostree and systemd under macos

### CODE REVIEW

File: src/aktualizr_secondary/CMakeLists.txt [hunk 1]
```diff
@@ -48,9 +48,12 @@ target_link_libraries(aktualizr-secondary
     ${LIBDPKG_LIBRARIES}
     ${LIBP11_LIBRARIES}
     ${SYSTEMD_LIBRARY}
-    gcov
     )
 
+if(BUILD_WITH_CODE_COVERAGE)
+    target_link_libraries(aktualizr_secondary gcov)
```
Comment (lbonn) [inline src/aktualizr_secondary/CMakeLists.txt]: Typo here: `aktualizr-secondary` (executable names are hyphen-separated)

File: src/aktualizr_info/CMakeLists.txt [hunk 2]
```diff
@@ -11,8 +11,12 @@ target_link_libraries(aktualizr-info aktualizr_static_lib
   ${GLIB2_LIBRARIES}
   ${LibArchive_LIBRARIES}
   ${LIBOSTREE_LIBRARIES}
-  ${SQLITE3_LIBRARIES}
-  gcov)
+  ${SQLITE3_LIBRARIES})
+
+
+if(BUILD_WITH_CODE_COVERAGE)
+    target_link_libraries(aktualizr_info gcov)
```
Comment (lbonn) [inline src/aktualizr_info/CMakeLists.txt]: Here as well: `aktualizr-info`

File: src/cert_provider/main.cc [hunk 3]
```diff
@@ -230,37 +230,35 @@ bool generate_and_sign(const std::string& cacert_path, const std::string& capkey
...
     return false;
   }
-  ret = PEM_write_X509(cert_file, certificate.get());
-  fclose(cert_file);
+  ret = PEM_write_bio_X509(cert_file, certificate.get());
   if (!ret) {
     std::cerr << "PEM_write_X509" << std::endl;
     return false;
   }
+  auto cert_len = BIO_get_mem_data(cert_file, cert_buf);
```
[the remark this answers is not in the document]
Comment (cajun-rat) [reply src/cert_provider/main.cc]: Wow, openssl *is* special!

[the remark this answers is not in the document]
Comment (patrickvacek) [reply src/cert_provider/main.cc]: Brilliant design. I really wish we could do without macro magic at this point.

[the remark this answers is not in the document]
Comment (koshelev) [reply src/cert_provider/main.cc]: Good catch! Second argument should be a pointer to a pointer` long BIO_get_mem_data(BIO *b, char **pp)`

**Comment by lbonn** [comment]
All this manual memory management makes me anxious. I'm gonna work on using `StructGuard` everywhere when we work with openssl objects.

File: tests/crypto_test.cc [hunk 4]
```diff
@@ -233,7 +233,7 @@ TEST(crypto, parsep12_FAIL) {
     EXPECT_TRUE(false) << " could not open tests/test_data/priv.key";
   }
   bool result = Crypto::parseP12(p12src, "", &pkey, &cert, &ca);
-  fclose(bad_p12file);
+  BIO_free(p12src);
```
Comment (koshelev) [inline tests/crypto_test.cc]: Ouch... Good catch, thanks!

File: src/socketgateway.cc [hunk 5]
```diff
@@ -8,6 +8,10 @@
 
 #include "logging.h"
 
+#ifndef MSG_NOSIGNAL
```
Comment (patriotyk) [inline src/socketgateway.cc]: Why do we need this defines?

Comment (lbonn) [reply src/socketgateway.cc]: Used to not get a `SIGPIPE` after a broken pipe during `send()`. Note that we also incorrectly use it for `recv` but according to the man page it doesn't apply there. We'd probably want to use it for our secondary ipc implem as well...

It's not available on OSX: https://nwat.xyz/blog/2014/01/16/porting-msg_more-and-msg_nosigpipe-to-osx/.

File: src/utilities/crypto.cc [hunk 6]
```diff
@@ -175,70 +175,69 @@ bool Crypto::VerifySignature(const PublicKey &public_key, const std::string &sig
...
+  ca_len = BIO_get_mem_data(ca_sink, &ca_buf);
   *out_ca = std::string(ca_buf, ca_len);
-  free(ca_buf);
+  BIO_free(ca_sink);
 
-  fclose(cert_file);
+  cert_len = BIO_get_mem_data(cert_sink, &cert_buf);
   *out_cert = std::string(cert_buf, cert_len);
-  free(cert_buf);
+  BIO_free(cert_sink);
```
Comment (cajun-rat) [inline src/utilities/crypto.cc]: We should think about a openssl BIO wrapper class with methods to convert to std::string etc, but I think @lbonn is already looking into that.

Comment (lbonn) [reply src/utilities/crypto.cc]: Right now, the wrappers are just unique_ptr and do not provide any additional helper methods. But we can discuss suggestions in #717

**Pull request merged**